Sources/NIOCore/IPProtocol.swift

```diff
@@ -0,0 +1,338 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// In the Internet Protocol version 4 (IPv4) [RFC791] there is a field
+/// called "Protocol" to identify the next level protocol.  This is an 8
+/// bit field.  In Internet Protocol version 6 (IPv6) [RFC8200], this field
+/// is called the "Next Header" field.
+public struct NIOIPProtocol: RawRepresentable, Hashable {
+    public typealias RawValue = UInt8
+    public var rawValue: RawValue
+    
+    @inlinable
+    public init(rawValue: RawValue) {
```

Comment:
Can we add an `init(_ rawValue: Int)` as well? It's useful to be able to construct this from a general `Int`.

Reply:
I think so. It remains helpful for cases where you can easily get hold of an `Int`, and represents good practice in the Swift ecosystem (where `Int` is the currency numerical type). The crash at runtime is a red herring: if the user has an `Int`, we just force them to write the crash in _their_ code instead of ours. The type very clearly communicates that it can only hold a `UInt8`, so there's no ambiguity, and we can safely save the user the hassle of writing that code.

Reply:
Done in 35cb11f9b16375ccbedccd4907afc8f2af668b6b